dao/src/main/java/com/epam/esm/repository/GiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,222 @@
+package com.epam.esm.repository;
+
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.GiftsTags;
+import com.epam.esm.entity.SearchTags;
+import com.epam.esm.entity.Tag;
+import lombok.RequiredArgsConstructor;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.springframework.stereotype.Repository;
+
+import javax.persistence.TypedQuery;
+import javax.persistence.criteria.*;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+
+@Repository
+@RequiredArgsConstructor
+public class GiftCertificateDaoImpl implements GiftCertificateDao {
+
+	private final SessionFactory sessionFactory;
+
+	@Override
+	public void save(GiftCertificate gc) {
+		Session session = sessionFactory.openSession();
+		session.beginTransaction();
+		session.saveOrUpdate(gc);
+		session.getTransaction().commit();
+		session.close();
+	}
+
+	@Override
+	public GiftCertificate findById(Integer id) {
+		Session session = sessionFactory.openSession();
+		CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+		CriteriaQuery<GiftCertificate> criteriaQuery = criteriaBuilder.createQuery(GiftCertificate.class);
+		Root<GiftCertificate> root = criteriaQuery.from(GiftCertificate.class);
+		root.fetch("tags", JoinType.LEFT);
+		criteriaQuery.select(root).where(criteriaBuilder.equal(root.get("id"), id));
```

Comment:
i'm not sure, but probably `.select(root)` can be omitted as entity is already specified in criteria